### PR TITLE
Added Network Table / Node refactoring

### PR DESCRIPTION
**Pull request by turcato-niccolo**

### CODE REVIEW

File: distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/DistributedNetworkNode.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+
+    /**
+     * @param toHash  the String (a key) of which generate hashcode
+     * @param numBits number of bits that the hash code will be constituted of, must be multiple of 64 and >0
+     * @return the bitSet containing the hash of the peer's address, bitSet's length is a multiple of 64
+     * @throws IllegalArgumentException if peer isn't valid or the numBit isn't multiple of 64
+     */
+    public static BitSet hash(@NonNull String toHash, int numBits) {
```
Comment (bortoletti-giorgia) [inline distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/DistributedNetworkNode.java]: In this method you can use some encryption algorithm as SHA-1 for a key of 160 bits or MD5 for a key of 128 bit.
Like this:
```
try{
            MessageDigest md = MessageDigest.getInstance("SHA-1");
            md.update(toHash.getBytes());
            byte[] digest = md.digest();

            id = fromByteArray(digest);
}catch(NoSuchAlgorithmException e){}
```

---and---
```
/**
     * Returns a bitset containing the values in bytes.
     * @param bytes to convert in BitSet object
     */
    public static BitSet fromByteArray(byte[] bytes) {
      BitSet bitSet = new BitSet();
      for (int i = 0; i < bytes.length * 8; i++) {
        if ((bytes[bytes.length - i / 8 - 1] & (1 << (i % 8))) > 0) {
        	bitSet.set(i);
        }
      }
      return bitSet;
    }
```

File: distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/NetworkDictionaryWithOwnership.java [hunk 2]
```diff
@@ -20,15 +20,15 @@
 
     /**
      * Constructor, initializes the object and defines the backup database
+     *
```
Comment (bortoletti-giorgia) [inline distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/NetworkDictionaryWithOwnership.java:23]: In this class I would write Map<Peer, Node[]>.

File: distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/RoutingTable.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    /**
+     * @param node DistributedNetworkNode of the distributed Network
+     * @return the index (between 0 and N -1) of the bucket containing (or that should contain) the given DistributedNetworkNode (Resource or Peer)
+     */
+    public abstract int getLocation(Node node);
+
+    /**
+     * @return true if the node has been added, false otherwise
+     */
+    public abstract boolean Add(Node node);
```
Comment (bortoletti-giorgia) [inline distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/RoutingTable.java:46]: It is a minimal correction but each method should start with a lowercase letter.

File: distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/KBucket.java [hunk 4]
```diff
@@ -1,56 +1,64 @@
...
- */
-public class KBucket extends Bucket<PeerNode> {
+
+import android.content.Context;
+
+import java.security.acl.NotOwnerException;
+
+public class KBucket extends Bucket<Node> {
+
     private int dimension;
```
Comment (bortoletti-giorgia) [inline distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/KBucket.java:4]: "dimension" is present also in Bucket<Node> so it isn't necessary to insert it and just add a getDimension() on Bucket<Node> and on the constructor of KBucket invoke "super(dim)".

**Comment by ricdezen** [comment]
Sorry, I've been working on this for a while and I can't seem to find a good reason to define our bucket ourselves, we have no iterator, no multiple adding and removal, we simply created a class that does less than what an ArrayList would already do. Also the Nodes are very complicated, what do we use in the network? PeerNode and ResourceNode or ReplicatedNetworkNode?
I'll leave the Buckets as they are nd focus on the Nodes

**Comment by RicDeZen** [comment]
Actually I'll jump straight to adding requests and such, I can't expect to overwrite everything everyone does

**Comment by turcato-niccolo** [comment]
> Sorry, I've been working on this for a while and I can't seem to find a good reason to define our bucket ourselves, we have no iterator, no multiple adding and removal, we simply created a class that does less than what an ArrayList would already do. 

Doing more isn't the point, for example in the routing table we need containers that do just a set of operations, for this, I've written the abstract Bucket. KBucket is a realization of that, and I see it's not complete of course but it's not useless. If we need it to be iterable, we will implement it.

> Also the Nodes are very complicated, what do we use in the network? PeerNode and ResourceNode or ReplicatedNetworkNode?

Nodes are a little complicated but if you want to manage BitSets you need something like this, as Bitsets are not of fixed lenghts, and we shoudn't accept any kind of Bitsets, as you wouldn't know which address space they belong. It's a tradeoff for having a handy binary key.
Moreover if you have seen, with DistributedNetworkNode you can calculate distances between keys, which is handy if you want to sort them.
DistributedNetworkNode is what we would use as Keys in the distributed network, both PeerNode and ResourceNode return an istance of DistributedNetworkNode with method getAddress.
It was the simplest way, that I've thought of, to have a generalized Ojbect for the addresses/keys of the distributed network and two objects that extends the abstract classes Peer< T > and Resource<K,V>. So that we could reuse objects defined on them, like maybe the dictionary used also in NetworkManager.

> I'll leave the Buckets as they are nd focus on the Nodes

Also, a small note: both Buckets and Nodes were added last week: mo' me lo dici?😂😂

File: distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/DistributedNetworkNode.java (hunk 1 above)
Comment (turcato-niccolo) [reply distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/DistributedNetworkNode.java]: This is much better! Thanks!

**Pull request merged**